apis/vshn/v1/dbaas_vshn_Codey.go

```diff
@@ -0,0 +1,181 @@
+//go:generate yq -i e ../../generated/vshn.appcat.vshn.io_codeys.yaml --expression "with(.spec.versions[]; .schema.openAPIV3Schema.properties.spec.properties.parameters.properties.size.default={})"
+//go:generate yq -i e ../../generated/vshn.appcat.vshn.io_codeys.yaml --expression "with(.spec.versions[]; .schema.openAPIV3Schema.properties.spec.properties.parameters.properties.service.default={})"
+
+// +kubebuilder:object:root=true
+
+// Codey is the API for creating Codey instances.
+type Codey struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of a Codey.
+	Spec CodeySpec `json:"spec"`
+
+	// Status reflects the observed state of a Codey.
+	Status CodeyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:generate=true
+// +kubebuilder:object:root=true
+type CodeyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Codey `json:"items,omitempty"`
+}
+
+// CodeySpec defines the desired state of a Codey.
+type CodeySpec struct {
+	// Parameters are the configurable fields of a Codey.
+	Parameters CodeyParameters `json:"parameters,omitempty"`
+
+	// WriteConnectionSecretToRef references a secret to which the connection details will be written.
+	WriteConnectionSecretToRef LocalObjectReference `json:"writeConnectionSecretToRef,omitempty"`
+}
+
+// CodeyParameters are the configurable fields of a Codey.
+type CodeyParameters struct {
+	// Service contains Codey DBaaS specific properties
+	Service CodeyServiceSpec `json:"service,omitempty"`
+
+	// Size contains settings to control the sizing of a service.
+	Size VSHNSizeSpec `json:"size,omitempty"`
+}
+
+// CodeyServiceSpec contains Codey DBaaS specific properties
+type CodeyServiceSpec struct {
+
+	// Version contains supported version of Codey.
+	// Multiple versions are supported.
+	MajorVersion string `json:"majorVersion,omitempty"`
+
+	// Codeysettings contains additional Codey settings.
+	AdminEmail string `json:"adminEmail,omitempty"`
+}
+
+// CodeySizeSpec contains settings to control the sizing of a service.
+type CodeySizeSpec struct {
+
+	// CPURequests defines the requests amount of Kubernetes CPUs for an instance.
+	CPURequests string `json:"cpuRequests,omitempty"`
+
+	// CPULimits defines the limits amount of Kubernetes CPUs for an instance.
+	CPULimits string `json:"cpuLimits,omitempty"`
+
+	// MemoryRequests defines the requests amount of memory in units of bytes for an instance.
+	MemoryRequests string `json:"memoryRequests,omitempty"`
+
+	// MemoryLimits defines the limits amount of memory in units of bytes for an instance.
+	MemoryLimits string `json:"memoryLimits,omitempty"`
+
+	// Disk defines the amount of disk space for an instance.
+	Disk string `json:"disk,omitempty"`
+
+	// Plan is the name of the resource plan that defines the compute resources.
```

Comment:
We should probably make plan an enum, so users can't choose invalid plans.

A default would be nice as well.